dvc/scm/git/__init__.py

```diff
@@ -244,8 +244,18 @@ def list_branches(self):
     def list_tags(self):
         return [t.name for t in self.repo.tags]
 
-    def list_all_commits(self):
-        return [c.hexsha for c in self.repo.iter_commits("--all")]
```

Comment:
This was a misuse of `iter_commits`, where passing `--all` to it worked accidentally because it was passing it as `rev` to `rev-list`. New implementation uses `rev-list` directly, which should be significantly faster (no benches yet).

Reply:
Also, previous implementation was silently keeping `refs/remotes` and `refs/stash` as well, which might or might not be what we want, considering that we don't currently have any means of deleting cache for those.

Reply:
So it was previously simply "keep anything, which is referenced from git". Now it's more complicated )